app/Http/Controllers/Api/Web/AlbumsController.php

```diff
@@ -83,6 +86,53 @@ public function getShow($id)
         ], 200);
     }
 
+    public function getCachedAlbum($id, $format)
```

Comment:
There aren't hard-and-fast rules for this but I go by the notion that the logic in a controller action should exist primarily for determining what response to return and, in some cases, mapping variables to a view or data structure.

Logic that details with inner details of how data is implemented seems more conceptually appropriate for the model classes. In this particular case, following the first two checks, there are three possible outcomes:

- some track in the requested album does not have a track file in the requested format -> return 404
- one or more of the album's tracks needs to be encoded in the requested format before the download is ready -> kick off the encoding and return the null URL
- all of the album's track files are ready to go -> return download URL

Where the controller is concerned, you can implement methods on the `Album` class that check for each of these conditions, return true or false, and let you write simple if statements (and contain the messier logic involved with determining these conditions in dedicated functions).